### ISSUE TITLE
Pi4 Menu from CLi is messed. X11 is ok

### ISSUE THREAD

**Issue by 2play** [dhewm/dhewm3#368]
Hi to all. I m running doom 3 and d3re on x11 without problem
from cli i can run but menu has the DOOM 3 logo in front ot the planet all messed up.
I tried +set cl_renderer opengl1 or oepngl2 but no difference
From X11 is ok.

Here are the examples
CLi
![printscreen04_12_2021--h21-m50-s32](https://user-images.githubusercontent.com/30235736/114464350-ee22b480-9be5-11eb-8f3b-0588f8efca21.jpg)

![printscreen04_12_2021--h21-m51-s39](https://user-images.githubusercontent.com/30235736/114464387-fda1fd80-9be5-11eb-9b44-70a85f506d32.jpg)

X11
![printscreen04_12_2021--h22-m26-s13](https://user-images.githubusercontent.com/30235736/114464772-7bfe9f80-9be6-11eb-9484-e94d2aaf5681.jpg)

**Comment by DanielGibson** [comment]
looks the same as #348 
no idea why it happens though

**Comment by 2play** [comment]
I think it might be renderer related. any other options i can set ?

**Comment by DanielGibson** [comment]
Just for reference, `+set cl_renderer opengl1` doesn't make sense, dhewm3 has no such cvar and only has one renderer.

One thing that you could try is setting `r_waylandcompat 1`

**Comment by 2play** [comment]
tested. fails with below

![printscreen04_12_2021--h23-m12-s07](https://user-images.githubusercontent.com/30235736/114469165-ec102400-9bec-11eb-8003-9d47560e3f0e.jpg)

**Comment by DanielGibson** [comment]
I don't see any failing there

**Comment by 2play** [comment]
it goes straight to console with above warnings.  :-(

**Comment by DanielGibson** [comment]
that's weird (note that the output is the same as in your earlier screenshot with the console visible)

**Comment by 2play** [comment]
you are right. on my first i entered the console to see if any error due to the messed logo
this time just went to full screen console

**Comment by DanielGibson** [comment]
and it can't be closed?

by the way, I just noticed that when I enable r_waylandcompat (on my X11 system with nvidia drivers) I get the same glitch

**Comment by DanielGibson** [comment]
can you somehow post the whole console output (it's also written to stdout)?

**Comment by 2play** [comment]
i tried to start with XINIT: all ok but exit stays to black screen with cursor (wont exit)
`xinit ~/games/doom3/dhewm3 r_waylandcompat 1 -- vt1 -keeptty  gives this (no gamma error)`
![printscreen04_12_2021--h23-m26-s18](https://user-images.githubusercontent.com/30235736/114470272-e6b3d900-9bee-11eb-96df-17c320473a21.jpg)

`xinit ~/games/doom3/dhewm3 -- vt1 -keeptty`
boots but jsut part of the sceen (im using 640x480 as anything higher is not playable on pi4) and exits smooth

**Comment by DanielGibson** [comment]
you gotta use `... dhewm3 +set r_waylandcompat 1`

do you compile dhewm3 yourself? if not, can you do that and do one little change:
In `neo/sys/glimp.cpp` around line 370, *after* 
```c++
	common->Printf("Using %d color bits, %d depth, %d stencil display\n", 
					channelcolorbits, tdepthbits, tstencilbits);
```
please add the following code:
```c++
	{
			int r, g, b, a, d, s;
			SDL_GL_GetAttribute(SDL_GL_RED_SIZE, &r);
			SDL_GL_GetAttribute(SDL_GL_GREEN_SIZE, &g);
			SDL_GL_GetAttribute(SDL_GL_BLUE_SIZE, &b);
			SDL_GL_GetAttribute(SDL_GL_ALPHA_SIZE, &a);
			SDL_GL_GetAttribute(SDL_GL_DEPTH_SIZE, &d);
			SDL_GL_GetAttribute(SDL_GL_STENCIL_SIZE, &s);

			common->Printf("Got %d stencil bits, %d depth bits, color bits: %d r %d g %d b %d a\n", s, d, r, g, b, a);
		}
```

then compile dhewm3, start it from console and afterwards post the whole output from stdout

**Comment by 2play** [comment]
i ll try tomorrow to compile. its late here so i was testing things :-)
tx for your time and replies. much appreciated

i tested`xinit ~/games/doom3/dhewm3 +set r_waylandcompat 1 -- vt1 -keeptty`
it gives interestingly this
![printscreen04_12_2021--h23-m34-s05](https://user-images.githubusercontent.com/30235736/114471188-7a39d980-9bf0-11eb-982a-c20bb4e23938.jpg)

**Comment by 2play** [comment]
also `XINIT: ~/games/doom3/dhewm3 +set r_waylandcompat 1`
gives 
![printscreen04_12_2021--h23-m39-s37](https://user-images.githubusercontent.com/30235736/114471352-bd944800-9bf0-11eb-8e1d-6e72ecaee090.jpg)

**Comment by DanielGibson** [comment]
yes I'll need to output from the patched dhewm3 (most probably with `r_waylandcompat 0`) to make a better guess what goes wrong.

you should be able to save the output to a text file by starting (the patched) dhewm3 with `./dhewm3 > d3log.txt`

**Comment by 2play** [comment]
after killing this
![image](https://user-images.githubusercontent.com/30235736/114471468-fcc29900-9bf0-11eb-8868-7f6895d51ad0.png)

I retest without `XINIT: ~/games/doom3/dhewm3`
and i have 
![printscreen04_12_2021--h23-m45-s14](https://user-images.githubusercontent.com/30235736/114471834-883c2a00-9bf1-11eb-9357-85a6832b1cef.jpg)

**Comment by 2play** [comment]
last is the best option but wont exit unless killed :-(

**Comment by 2play** [comment]
```
dhewm3 1.5.2pre.1305 linux-armv7l Apr 10 2021 23:46:53 using SDL v2.0.10
terminal support enabled ( use +set in_tty 0 to disable )
pid: 31187
1568 MB System Memory
found interface lo - loopback
found interface wlan0 - 192.168.1.106/255.255.255.0
WARNING: base path '/usr/local/share/dhewm3' does not exist
WARNING: using path of executable: /home/pi/games/doom3
----- Initializing File System -----
Loaded pk4 /home/pi/games/doom3/base/pak000.pk4 with checksum 0x28d208f1
Loaded pk4 /home/pi/games/doom3/base/pak001.pk4 with checksum 0x40244be0
Loaded pk4 /home/pi/games/doom3/base/pak002.pk4 with checksum 0xc51ecdcd
Loaded pk4 /home/pi/games/doom3/base/pak003.pk4 with checksum 0xcd79d028
Loaded pk4 /home/pi/games/doom3/base/pak004.pk4 with checksum 0x765e4f8b
Loaded pk4 /home/pi/games/doom3/base/pak005.pk4 with checksum 0x8ffc3621
Loaded pk4 /home/pi/games/doom3/base/pak006.pk4 with checksum 0x95b65ab
Loaded pk4 /home/pi/games/doom3/base/pak007.pk4 with checksum 0x666bdb3c
Loaded pk4 /home/pi/games/doom3/base/pak008.pk4 with checksum 0x23ae5993
Current search path:
/home/pi/.config/dhewm3/base
/home/pi/.local/share/dhewm3/base
/home/pi/games/doom3/base
/home/pi/games/doom3/base/pak008.pk4 (3 files)
/home/pi/games/doom3/base/pak007.pk4 (38 files)
/home/pi/games/doom3/base/pak006.pk4 (48 files)
/home/pi/games/doom3/base/pak005.pk4 (63 files)
/home/pi/games/doom3/base/pak004.pk4 (5137 files)
/home/pi/games/doom3/base/pak003.pk4 (4676 files)
/home/pi/games/doom3/base/pak002.pk4 (6120 files)
/home/pi/games/doom3/base/pak001.pk4 (8972 files)
/home/pi/games/doom3/base/pak000.pk4 (2698 files)
Addon pk4s:
----- Initializing Decls -----
5206 strings read from strings/english.lang
Couldn't open journal files
execing editor.cfg
execing default.cfg
execing dhewm.cfg
couldn't exec autoexec.cfg
5206 strings read from strings/english.lang
----- Initializing OpenAL -----
Setup OpenAL device and context
OpenAL: found device 'Built-in Audio Digital Stereo'
OpenAL vendor: OpenAL Community
OpenAL renderer: OpenAL Soft
OpenAL version: 1.1 ALSOFT 1.19.1
OpenAL: found extensions for resetting disconnected devices
OpenAL: found EFX extension
OpenAL: found 256 hardware voices
----- Initializing OpenGL -----
Initializing OpenGL subsystem
Using 8 color bits, 24 depth, 8 stencil display
OpenGL vendor: Broadcom
OpenGL renderer: V3D 4.2
OpenGL version: 2.1 Mesa 19.3.2
...using GL_ARB_multitexture
...using GL_ARB_texture_env_combine
...using GL_ARB_texture_cube_map
...using GL_ARB_texture_env_dot3
...using GL_ARB_texture_env_add
...using GL_ARB_texture_non_power_of_two
...using GL_ARB_texture_compression
X..GL_EXT_texture_compression_s3tc not found
X..GL_EXT_texture_filter_anisotropic not found
...using GL_1.4_texture_lod_bias
X..GL_EXT_shared_texture_palette not found
...using GL_EXT_texture3D
...using GL_EXT_stencil_wrap
...using GL_EXT_stencil_two_side
... got GL2.0+ glStencilOpSeparate()
...using GL_ARB_vertex_buffer_object
...using GL_ARB_vertex_program
...using GL_ARB_fragment_program
X..EXT_depth_bounds_test not found
ARB2 renderer: Available.
----- R_ReloadARBPrograms -----
glprogs/test.vfp
glprogs/test.vfp
glprogs/interaction.vfp
glprogs/interaction.vfp
glprogs/bumpyEnvironment.vfp
glprogs/bumpyEnvironment.vfp
glprogs/ambientLight.vfp
glprogs/ambientLight.vfp
glprogs/shadow.vp
glprogs/environment.vfp
glprogs/environment.vfp
glprogs/arbVP_glasswarp.txt: File not found
glprogs/arbFP_glasswarp.txt: File not found
using ARB_vertex_buffer_object memory
using ARB2 renderSystem
WARNING: Couldn't set gamma ramp: That operation is not supported
WARNING: Couldn't set gamma ramp: That operation is not supported
loaded game library '/home/pi/games/doom3/base.so'.
----- Initializing Game -----
gamename: baseDOOM-1
gamedate: Apr 10 2021
Initializing event system
...473 event definitions
Initializing class hierarchy
...142 classes, 382184 bytes for event callbacks
Initializing scripts
Compiled 'weapon_pistol::Raise': 1466 ms
----- Compile stats -----
Memory usage:
     Strings: 79, 12592 bytes
  Statements: 67875, 1357500 bytes
   Functions: 2109, 250532 bytes
   Variables: 147376 bytes
    Mem used: 2579384 bytes
 Static data: 2477552 bytes
   Allocated: 3484640 bytes
 Thread size: 13212 bytes
...6 aas types
----- Initializing Session -----
----- Warnings -----
during dhewm 3 initialization...
WARNING: base path '/usr/local/share/dhewm3' does not exist
WARNING: Couldn't set gamma ramp: That operation is not supported
WARNING: using path of executable: /home/pi/games/doom3
3 warnings
]exit
----- Game Map Shutdown -----
Shutting down sound hardware
idRenderSystem::Shutdown()
Shutting down OpenGL subsystem
----- Game Shutdown -----
----- Game Map Shutdown -----
Shutdown event system
shutdown terminal support

```

**Comment by DanielGibson** [comment]
Yeah, that's the output I'll need, just with my patch applied so we'll get to know the actually used color/alpha bits.
No problem if you do that tomorrow, until then I can prepare a git branch for you with the patch already applied so building will be easier for you (I'll post the link here later)

**Comment by DanielGibson** [comment]
The branch is here: https://github.com/dhewm/dhewm3/tree/debug-gfxglitch
If you're not familiar with git, you can also download it as a [.zip](https://github.com/dhewm/dhewm3/archive/refs/heads/debug-gfxglitch.zip) or [.tar.gz](https://github.com/dhewm/dhewm3/archive/refs/heads/debug-gfxglitch.tar.gz).

Compiling instructions are in the [README](https://github.com/dhewm/dhewm3/blob/master/README.md#compiling).
Note that you won't need libjpeg, libogg or libvorbis(file), because the debug-gfxglitch branch already has the changes from #363  which replace those libs with header-only libs included in the dhewm3 source.

**Comment by 2play** [comment]
hi @DanielGibson 
I have compiled branch and used the 3 new files. 
from CLi DOOM logo is still messed. 
here is the log you requested

```dhewm3 1.5.2pre.1305 linux-armv7l Apr 13 2021 09:03:55 using SDL v2.0.10
terminal support enabled ( use +set in_tty 0 to disable )
pid: 9661
1568 MB System Memory
found interface lo - loopback
found interface wlan0 - 192.168.1.106/255.255.255.0
WARNING: base path '/usr/local/share/dhewm3' does not exist
WARNING: using path of executable: /home/pi/games/doom3
----- Initializing File System -----
Loaded pk4 /home/pi/games/doom3/base/pak000.pk4 with checksum 0x28d208f1
Loaded pk4 /home/pi/games/doom3/base/pak001.pk4 with checksum 0x40244be0
Loaded pk4 /home/pi/games/doom3/base/pak002.pk4 with checksum 0xc51ecdcd
Loaded pk4 /home/pi/games/doom3/base/pak003.pk4 with checksum 0xcd79d028
Loaded pk4 /home/pi/games/doom3/base/pak004.pk4 with checksum 0x765e4f8b
Loaded pk4 /home/pi/games/doom3/base/pak005.pk4 with checksum 0x8ffc3621
Loaded pk4 /home/pi/games/doom3/base/pak006.pk4 with checksum 0x95b65ab
Loaded pk4 /home/pi/games/doom3/base/pak007.pk4 with checksum 0x666bdb3c
Loaded pk4 /home/pi/games/doom3/base/pak008.pk4 with checksum 0x23ae5993
Current search path:
/home/pi/.config/dhewm3/base
/home/pi/.local/share/dhewm3/base
/home/pi/games/doom3/base
/home/pi/games/doom3/base/pak008.pk4 (3 files)
/home/pi/games/doom3/base/pak007.pk4 (38 files)
/home/pi/games/doom3/base/pak006.pk4 (48 files)
/home/pi/games/doom3/base/pak005.pk4 (63 files)
/home/pi/games/doom3/base/pak004.pk4 (5137 files)
/home/pi/games/doom3/base/pak003.pk4 (4676 files)
/home/pi/games/doom3/base/pak002.pk4 (6120 files)
/home/pi/games/doom3/base/pak001.pk4 (8972 files)
/home/pi/games/doom3/base/pak000.pk4 (2698 files)
Addon pk4s:
----- Initializing Decls -----
5206 strings read from strings/english.lang
Couldn't open journal files
execing editor.cfg
execing default.cfg
execing dhewm.cfg
couldn't exec autoexec.cfg
5206 strings read from strings/english.lang
----- Initializing OpenAL -----
Setup OpenAL device and context
OpenAL: found device 'Built-in Audio Digital Stereo'
OpenAL vendor: OpenAL Community
OpenAL renderer: OpenAL Soft
OpenAL version: 1.1 ALSOFT 1.19.1
OpenAL: found extensions for resetting disconnected devices
OpenAL: found EFX extension
OpenAL: found 256 hardware voices
----- Initializing OpenGL -----
Initializing OpenGL subsystem
Requested 8 color bits per chan, 8 alpha 24 depth, 8 stencil
Got 8 stencil bits, 24 depth bits, color bits: r8 g8 b8 a8
OpenGL vendor: Broadcom
OpenGL renderer: V3D 4.2
OpenGL version: 2.1 Mesa 19.3.2
...using GL_ARB_multitexture
...using GL_ARB_texture_env_combine
...using GL_ARB_texture_cube_map
...using GL_ARB_texture_env_dot3
...using GL_ARB_texture_env_add
...using GL_ARB_texture_non_power_of_two
...using GL_ARB_texture_compression
X..GL_EXT_texture_compression_s3tc not found
X..GL_EXT_texture_filter_anisotropic not found
...using GL_1.4_texture_lod_bias
X..GL_EXT_shared_texture_palette not found
...using GL_EXT_texture3D
...using GL_EXT_stencil_wrap
...using GL_EXT_stencil_two_side
... got GL2.0+ glStencilOpSeparate()
...using GL_ARB_vertex_buffer_object
...using GL_ARB_vertex_program
...using GL_ARB_fragment_program
X..EXT_depth_bounds_test not found
ARB2 renderer: Available.
----- R_ReloadARBPrograms -----
glprogs/test.vfp
glprogs/test.vfp
glprogs/interaction.vfp
glprogs/interaction.vfp
glprogs/bumpyEnvironment.vfp
glprogs/bumpyEnvironment.vfp
glprogs/ambientLight.vfp
glprogs/ambientLight.vfp
glprogs/shadow.vp
glprogs/environment.vfp
glprogs/environment.vfp
glprogs/arbVP_glasswarp.txt: File not found
glprogs/arbFP_glasswarp.txt: File not found
using ARB_vertex_buffer_object memory
using ARB2 renderSystem
WARNING: Couldn't set gamma ramp: That operation is not supported
WARNING: Couldn't set gamma ramp: That operation is not supported
loaded game library '/home/pi/games/doom3/base.so'.
----- Initializing Game -----
gamename: baseDOOM-1
gamedate: Apr 13 2021
Initializing event system
...473 event definitions
Initializing class hierarchy
...142 classes, 382184 bytes for event callbacks
Initializing scripts
Compiled 'weapon_pistol::Raise': 1392 ms
----- Compile stats -----
Memory usage:
     Strings: 79, 12592 bytes
  Statements: 67875, 1357500 bytes
   Functions: 2109, 250532 bytes
   Variables: 147376 bytes
    Mem used: 2579384 bytes
 Static data: 2477552 bytes
   Allocated: 3484640 bytes
 Thread size: 13212 bytes
...6 aas types
----- Initializing Session -----
----- Warnings -----
during dhewm 3 initialization...
WARNING: base path '/usr/local/share/dhewm3' does not exist
WARNING: Couldn't set gamma ramp: That operation is not supported
WARNING: using path of executable: /home/pi/games/doom3
3 warnings
]quitt
Unknown command 'quitt'
]quit
----- Game Map Shutdown -----
Shutting down sound hardware
idRenderSystem::Shutdown()
Shutting down OpenGL subsystem
----- Game Shutdown -----
----- Game Map Shutdown -----
Shutdown event system
shutdown terminal support
```

**Comment by DanielGibson** [comment]
could you also post the output for when started from X11 (when the logo looks fine)?

**Comment by 2play** [comment]
Here you are...
[d3logx11.txt](https://github.com/dhewm/dhewm3/files/6304997/d3logx11.txt)

**Comment by DanielGibson** [comment]
Thanks!
Looks exactly the same, except setting the gamma ramp is only supported with X11.
Can you tell me how you start the game from "CLI"? Do you just run `./dhewm3` from tty1 or whatever, without X11 running?

**Comment by 2play** [comment]
welcome!
yes by calling the binary or from exiting ES to CLi or by a sh starting it 
ALSO i found the solution to my other problem when using ti start it with XINIT: and when exit gets stuck in the terminal
I noticed the comment on the log file ` +set in_tty 0` and when i start it like that then it exits fine plus logo is working ok since its not plain binary start out of X11 

At least i now have a way to start from ES have no glitches and exit smooth without terminal hung 

@DanielGibson Let me know if you want me to do further testing

**Comment by DanielGibson** [comment]
Another question: What do you run on your RPi4? Raspberry Pi OS (formerly Raspbian)?

**Comment by 2play** [comment]
Yes official latest updates. 
Clean Official image, manual install of RPie, 
Linux 5.10.17-v7l+ armv7l GNU/Linux - Raspbian GNU/Linux 10 (buster)
 sdl2-config --version 2.0.15
vulkaninfo | grep deviceName deviceName     = V3D 4.2
latest mesa  compiled as `-Dplatforms=x11 -Dvulkan-drivers=broadcom -Ddri-drivers= -Dgallium-drivers=v3d,kmsro,vc4 -Dbuildtype=debug -Dprefix=/usr build` not release on the host im testing.
I plan to replace with release so it can upgrade the 19.3.2 on Raspi OS

**Comment by DanielGibson** [comment]
> sdl2-config --version 2.0.15
that's weird, because
1. my raspberry pi os has 2.0.9
2. the log you posted says `dhewm3 1.5.2pre.1305 linux-armv7l Apr 13 2021 09:03:55 using SDL v2.0.10`

(I haven't run any tests on my RPi yet, I only just booted and updated it)

**Comment by 2play** [comment]
i know and apt policy libsdl2-2* reports

```Version table:
 *** 2.0.10+5rpi 100
        100 /var/lib/dpkg/status
     2.0.9+dfsg1-1+rpt1 500
        500 http://archive.raspberrypi.org/debian buster/main armhf Packages
     2.0.9+dfsg1-1 500
        500 http://raspbian.raspberrypi.org/raspbian buster/main armhf Packages```

**Comment by DanielGibson** [comment]
I can't get it to work at all from TTY on my RPi4 (4GB) with latest raspberry pi OS (buster, standard mesa), I always get
```
Initializing OpenGL subsystem
libEGL fatal: failed to initialize "current" system
```
and then a segfault.
I even built latest SDL git code myself, including kmsdrm support, and that didn't seem to help either.
(I'm not gonna build mesa.)

In X11 it works as expected.

**Comment by 2play** [comment]
@DanielGibson no worries friend.
im ok. Im running as I needed from ES or X11 with no artifacts. 
Pure CLi run has the artifacts but with xinit im good to run and exit as i need to.

**Comment by symbios24** [comment]
> @DanielGibson no worries friend.
> im ok. Im running as I needed from ES or X11 with no artifacts.
> Pure CLi run has the artifacts but with xinit im good to run and exit as i need to.
> A big thank you for all your time and guidance

2play did you have success running doom3 ftom cli? 
If yes can you tell me how you did it please?
I run it with startx ./dhewm3 it runs ok in a small window if i set it to Fullscreen it runs in a amall window in the top left down corner

**Comment by vrdaveb** [comment]
I was getting this with newer Mesa builds and a bunch of configurations on SDL 2.0.10, but it appears to be fixed in SDL 2.0.18. I was always getting RGBA8888, but there must have been a bug in the KMSDRM videodriver that is fixed now. If you end up building your SDL from source, be sure to pass ./configure --disable-video-rpi --enable-video-kmsdrm to avoid hard-to-catch errors.
